File: src/idmanager.cpp
#include "idmanager.hpp"

IdManager::IdManager(long max) : ids(max) {
  for (long i = 0; i < max; ++i)
    ids.push_back(i);
};

IdManager::~IdManager(){};

bool IdManager::add(long val, long &id) {
  if (ids.empty())
    return false;

  long new_id = ids[0];
  ids.pop_front();

  id_map[new_id] = val;
  id = new_id;
  return true;
}

bool IdManager::remove(long id, long &val) {
  map<long, long>::iterator pos = id_map.begin();
  while (pos != id_map.end()) {
    if (pos->first == id) {
      val = pos->second;
      ids.push_back(id);
      id_map.erase(pos);
      return true;
    }
    ++pos;
  }
  return false;
}

bool IdManager::remove_by_val(long val, long &id) {
  map<long, long>::iterator pos = id_map.begin();
  while (pos != id_map.end()) {
    if (pos->second == val) {
      id = pos->first;
      ids.push_back(id);
      id_map.erase(pos);
      return true;
    }
    ++pos;
  }
  return false;
}

bool IdManager::get(long id, long &val) {
  map<long, long>::iterator pos = id_map.find(id);
  if (pos == id_map.end())
    return false;

  val = pos->second;
  return true;
}

bool IdManager::get_id(long val, long &id) {
  map<long, long>::iterator pos = id_map.begin();
  while (pos != id_map.end()) {
    if (pos->second == val) {
      id = pos->first;
      return true;
    }
    ++pos;
  }
  return false;
}

void IdManager::iterate(void (*f)(long id, long val)) {
  map<long, long>::iterator pos = id_map.begin();
  while (pos != id_map.end()) {
    f(pos->first, pos->second);
    ++pos;
  }
}

int IdManager::size() {
    return ids.size();
}

File: rust/src/id_manager.rs
use std::collections::{HashMap, VecDeque};

pub struct IdManager {
    available_ids: VecDeque<i64>,
    id_map: HashMap<i64, i64>,
}

impl IdManager {
    pub fn new(max: i64) -> Self {
        let mut available_ids = VecDeque::with_capacity(max as usize);
        for i in 0..max {
            available_ids.push_back(i);
        }
        IdManager {
            available_ids,
            id_map: HashMap::new(),
        }
    }

    pub fn add(&mut self, val: i64) -> Option<i64> {
        if let Some(id) = self.available_ids.pop_front() {
            self.id_map.insert(id, val);
            Some(id)
        } else {
            None
        }
    }

    pub fn remove(&mut self, id: i64) -> Option<i64> {
        if let Some(val) = self.id_map.remove(&id) {
            self.available_ids.push_back(id);
            Some(val)
        } else {
            None
        }
    }

    pub fn remove_by_val(&mut self, val: i64) -> Option<i64> {
        let id = self.id_map.iter()
            .find(|&(_, &v)| v == val)
            .map(|(&k, _)| k);

        if let Some(id) = id {
            self.id_map.remove(&id);
            self.available_ids.push_back(id);
            Some(id)
        } else {
            None
        }
    }

    pub fn get(&self, id: i64) -> Option<i64> {
        self.id_map.get(&id).copied()
    }

    pub fn get_id(&self, val: i64) -> Option<i64> {
        self.id_map.iter()
            .find(|&(_, &v)| v == val)
            .map(|(&k, _)| k)
    }

    pub fn iterate<F>(&self, mut f: F)
    where
        F: FnMut(i64, i64),
    {
        for (&id, &val) in &self.id_map {
            f(id, val);
        }
    }

    pub fn size(&self) -> usize {
        self.available_ids.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_id_manager() {
        let mut manager = IdManager::new(3);
        assert_eq!(manager.size(), 3);

        let id1 = manager.add(10).unwrap();
        let id2 = manager.add(20).unwrap();
        let _id3 = manager.add(30).unwrap();
        assert!(manager.add(40).is_none());

        assert_eq!(manager.get(id1), Some(10));
        assert_eq!(manager.get_id(20), Some(id2));

        manager.remove(id2);
        assert_eq!(manager.size(), 1);
        let id4 = manager.add(40).unwrap();
        assert_eq!(id2, id4);

        manager.remove_by_val(10);
        assert_eq!(manager.get(id1), None);
    }
}
